### PR TITLE
Implement tower::Layer for tonic_web::Config

### LINKED ISSUE #1115: Remove `NamedService` bound for `tonic_web::enable`
## Bug Report

`tonic_web::enable` has a bound on `NamedService` used only for logging purposes.

### Version

```
➜ cargo tree | grep tonic
│   ├── tonic v0.8.0
│   ├── tonic v0.8.0 (*)
├── tonic v0.8.0 (*)
├── tonic-web v0.4.0
│   ├── tonic v0.8.0 (*)
└── tonic-build v0.8.0
├── tonic v0.8.0 (*)
```

### Crates

* `tonic-web`

### Description

This bound is not really needed, and requires to be implemented in cases such as when users manually implement the underlying service. Does it makes sense to remove it?

### LINKED ISSUE #1117: Implement `tower::Layer` for `tonic_web::Config`
## Feature Request

Per title, `tonic_web::Config` is doing pretty much the job of a `tower::Layer` (see `Config::enable` function). Would it be possible to implement the `tower::Layer` interface directly?

### Crates

* `tonic_web`

### Motivation

Better integration with tower based layers/services

### Proposal

Implement `tower::Layer` for `tonic_web::Config`

### Alternatives

Implement the layer manually now with a newtype

### PR DESCRIPTION
**Pull request by slinkydeveloper**
* Fix #1115
* Fix #1117

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
Perhaps one question @LucioFranco, should we document it somewhere in the rustdocs of the method `tonic_web::config` or of the struct `tonic_web::config::Config` that this trait is implemented? Or what rustdoc already generates in the trait implementation section is enough for you?

File: tonic-web/tests/integration/tests/grpc_web.rs [hunk 1]
```diff
@@ -74,14 +74,13 @@ async fn spawn(allowed_origin: &str) -> String {
...
-        .allow_origins(vec![allowed_origin])
-        .enable(TestServer::new(Svc));
+    let tonic_web_config = tonic_web::config().allow_origins(vec![allowed_origin]);
 
     let _ = tokio::spawn(async move {
         Server::builder()
             .accept_http1(true)
-            .add_service(svc)
+            .layer(tonic_web_config)
+            .add_service(TestServer::new(Svc))
```
Comment (LucioFranco) [inline tonic-web/tests/integration/tests/grpc_web.rs:83]: While I think the idea of adding a layer impl makes sense it feels like the code is actually a bit harder to understand at a glance. So I am inclined to continue to force users through the enabled api. I am curious to hear what @davidpdrsn and @neoeinstein thoughts might be.

**Comment by slinkydeveloper** [comment]
@LucioFranco in my case, I'm not using `tonic::Server` and I'm implementating manually the underlying service, reusing tonic only for decoding/encoding the grpc/grpc-web protocol. In this kind of use cases, supporting tower layers is really handy.

Somehow related to this (but I can open another issue to discuss it), I found out that the CORS code in tonic-web is doing more or less the same of https://docs.rs/tower-http/latest/tower_http/cors/index.html. I was wondering if it makes sense to just expose a GrpcWeb layer that transcodes grpc-web to grpc, and let the tower-http cors implementation handle the cors preflight. So one could either pick the "transcoding layer" manually, or can use `tonic_web::enable` to get the cors layer from tower-http composed with the transcoding layer. WDYT?

**Comment by slinkydeveloper** [comment]
@LucioFranco `to_layer` sounds good to me, although I have in mind something different (which involves breaking changes):

* `WebLayer` is exposed directly, it has no parameters (doesn't need them), has a `new` method, and it implements `Layer`, generating `GrpcWeb<S>` services
* `tonic_web::enable` composes `Cors` layer from tonic-http and `WebLayer`. If a user wants more control over the cors layer (like me), they can just instantiate manually `WebLayer` and compose it with whatever configured cors layer they want
* `Config` gets removed, there is no need for that anymore because all the fields are used to implement cors

WDYT?

**Comment by LucioFranco** [comment]
Yeah, that sounds like a great plan, smaller separate PRs will make my life much easier! This all sounds good to me. Thank you!

**Comment by slinkydeveloper** [comment]
@LucioFranco I think this PR as it is should be fine now. I'm working on the second PR to reimplement cors using `tower_http::cors`

**Review by LucioFranco — COMMENTED** [review]
LGTM just one question

File: tonic-web/src/layer.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+use super::{BoxBody, BoxError, Config, GrpcWeb};
+
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer implementing the grpc-web protocol.
+#[derive(Debug)]
+pub struct GrpcWebLayer {
+    _priv: (),
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply tonic-web/src/layer.rs:9]: I don't think so, as discussed on discord, in a followup PR i'm gonna remove `Config`

[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply tonic-web/src/layer.rs:9]: https://github.com/hyperium/tonic/issues/1122 here is a description of the next PR

**Review by LucioFranco — APPROVED** [review]
perfect! Thanks!

**Pull request merged**